pkg/progress/action.go

```diff
@@ -0,0 +1,215 @@
+package progress
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	crv1alpha1 "github.com/kanisterio/kanister/pkg/apis/cr/v1alpha1"
+	"github.com/kanisterio/kanister/pkg/client/clientset/versioned"
+	"github.com/kanisterio/kanister/pkg/field"
+	fn "github.com/kanisterio/kanister/pkg/function"
+	"github.com/kanisterio/kanister/pkg/log"
+	"github.com/kanisterio/kanister/pkg/reconcile"
+	"github.com/kanisterio/kanister/pkg/validate"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	progressPercentCompleted  = "100.00"
+	progressPercentStarted    = "10.00"
+	progressPercentNotStarted = "0.00"
+	weightNormal              = 1.0
+	weightHeavy               = 2.0
+	pollDuration              = time.Second * 2
+)
+
+var longRunningFuncs = []string{
+	fn.BackupDataFuncName,
+	fn.BackupDataAllFuncName,
+	fn.RestoreDataFuncName,
+	fn.RestoreDataAllFuncName,
+	fn.CopyVolumeDataFuncName,
+	fn.CreateRDSSnapshotFuncName,
+	fn.ExportRDSSnapshotToLocFuncName,
+	fn.RestoreRDSSnapshotFuncName,
+}
+
+// TrackActionsProgress tries to assess the progress of an actionSet by
+// watching the states of all the phases in its actions. It starts an infinite
+// loop, using a ticker to determine when to assess the phases. The function
+// returns when the provided context is either done or cancelled.
+// Caller should invoke this function in a non-main goroutine, to avoid
+// introducing any latencies on Kanister critical path.
+// If an error happens while attempting to update the actionSet, the failed
+// iteration will be skipped with no further retries, until the next tick.
+func TrackActionsProgress(
+	ctx context.Context,
+	client versioned.Interface,
+	actionSetName string,
+	namespace string,
+) error {
+	ticker := time.NewTicker(pollDuration)
+	defer ticker.Stop()
+
+	phaseWeights, totalWeight, err := calculatePhaseWeights(ctx, actionSetName, namespace, client)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+
+		case <-ticker.C:
+			actionSet, err := client.CrV1alpha1().ActionSets(namespace).Get(ctx, actionSetName, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			if actionSet.Status == nil {
+				continue
+			}
+
+			if err := updateActionsProgress(ctx, client, actionSet, phaseWeights, totalWeight, time.Now()); err != nil {
+				fields := field.M{
+					"actionSet":      actionSet.Name,
+					"nextUpdateTime": time.Now().Add(pollDuration),
+				}
+				log.Error().WithError(err).Print("failed to update phase progress", fields)
+				continue
```

Comment:
This may run into an infinite loop if the client fails to update action progress every iteration. Instead of continue, should we return error?

Reply:
The idea here is to ignore intermediate, flakey errors, due to e.g., API server or client-go throttling. The final progress should always be calculated based on the actionset status. If the actionset status completed, progress should always be 100% despite of intermediate errors.